### ISSUE TITLE
Constant-memory implementation of precision-recall related metrics

### ISSUE THREAD

**Issue by maximsch2** [PyTorchLightning/metrics#95]
## 🚀 Feature

Metrics that depend on precision-recall curve are currently implemented in a way that requires storing all of the predictions and labels in memory, making it's use impractical for large datasets or problem with large label spaces. We should support binning-based metrics implementation to solve this. Prototype is here: https://gist.github.com/maximsch2/2b55bab6deba629a5686258cb8152e53

### Alternatives

Don't do anything and be restricted in the scalability of metrics.

Other options for scaling is making it easier to keep metrics off-GPU.

A possible question is if we want to have both raw and binned implementations of the metrics.

### Additional context

Keras provides binning-based implementation by default: https://keras.io/api/metrics/classification_metrics/#auc-class

**Comment by SkafteNicki** [comment]
That would be a great improvement @maximsch2, and from the gist it already seems like you have nearly done the hard work. Please feel free to send a PR :]
